### PR TITLE
[WIP] Add KimiDeltaAttention

### PR DESCRIPTION
**Pull request by Waynezee**
### PR Category

User Experience

### PR Types

New features

### Description

添加KimiDeltaAttention

### 是否引起精度变化

否

### CODE REVIEW

**Comment by risemeup1111** [comment]
### Nyanpasu Review Board


| 序号 | 位置 | 优先级 | 状态 |
| - | - | - | - |
| 1 | [`src/paddlefleet/transformer/kimi_delta_attention.py`](https://github.com/PaddlePaddle/PaddleFleet/pull/1556#discussion_r3666516797) | P1 | 🚧 |
| 2 | [PR 标题](https://github.com/PaddlePaddle/PaddleFleet/pull/1556) | P3 | 🚧 |
| 3 | [`src/paddlefleet/transformer/kimi_delta_attention.py`](https://github.com/PaddlePaddle/PaddleFleet/pull/1556#discussion_r3666516486) | P1 | 🚧 |
| 4 | [`src/paddlefleet/transformer/kimi_delta_attention.py`](https://github.com/PaddlePaddle/PaddleFleet/pull/1556#discussion_r3681041608) | P1 | 🚧 |

<details>
<summary>点击展开 Review 规则</summary>
PR 评审规则：

- P0、P1 级别的评审必须提交新的 commit 进行修改；
- P2、P3 级别的评审可以通过评论进行修改。

状态说明：

| 已解决 | 无需修复 | 待修复 |
| - | - | - |
| ✅ | 🟡 | 🚧 |
</details>

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/transformer/kimi_delta_attention.py [hunk 1]
```diff
@@ -0,0 +1,751 @@
...
+    in_proj: LayerSpec | type = IdentityOp
+    f_a_proj: LayerSpec | type = IdentityOp
+    f_b_proj: LayerSpec | type = IdentityOp
+    g_a_proj: LayerSpec | type = IdentityOp
+    g_b_proj: LayerSpec | type = IdentityOp
+    out_norm: LayerSpec | type = IdentityOp
+    out_proj: LayerSpec | type = IdentityOp
+
+
+class KimiDeltaAttention(FleetLayer):
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/kimi_delta_attention.py:94]: P1 **优先级：P1**
**处理要求：请针对该评论修复并提交新的 commit。**

这个类新增了 `conv1d`、`dt_bias`、`A_log`、`f_*` / `g_*` 和 `out_proj` 等 TP 相关参数，但没有像 `GatedDeltaNet` 那样提供 `sharded_state_dict()`。这样做分布式 checkpoint 时，保存/恢复会缺少明确的 shard 规则，和现有 TP 层的 checkpoint 约定不一致。

建议补一个同风格的方法，至少把这些参数按各自分片规则纳入：
```python
def sharded_state_dict(self, structured_name_prefix: str = ""):
    # in_proj / f_b_proj / g_b_proj / out_proj 使用各自的 sharded_state_dict
    # conv1d、dt_bias、A_log 按 axis=0 shard
    # out_norm、f_a_proj、g_a_proj 保持 replicated
```

File: src/paddlefleet/transformer/kimi_delta_attention.py [hunk 2]
```diff
@@ -0,0 +1,751 @@
...
+            attention_mask: Unused, KDA is causal by construction (only fixed-length
+                input is supported, so there is no padding to mask out).
+            key_value_states: Key/value states (for cross attention, not supported).
+            attention_bias: Attention bias (unused).
+            packed_seq_params: Parameters used for THD format (not supported).
+
+        Returns:
+            Tuple of (output, output_bias).
+        """
+        if packed_seq_params is not None:
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/kimi_delta_attention.py:395]: P1 **优先级：P1**
**处理要求：请针对该评论修复并提交新的 commit。**

这里除了 `packed_seq_params` 之外，`attention_mask` / `attention_bias` / `attn_mask_startend_row_indices` 都被静默忽略了。只要这个层接到现有 transformer 路径上，带 padding 的 batch 就会把 padding token 送进递推状态，输出会偏掉但不会报错。

建议要么像 `GatedDeltaNet` 一样先构造并应用 padding mask，要么在收到非空 mask 时显式抛出 `NotImplementedError`，不要静默吞掉。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查，当前还有两处需要修正的实现问题，细节放在 inline review comments 里。

另外，PR 标题仍是 `[WIP]`，正文还是模板占位，且 `Check PR Description` 已失败。建议补齐实际变更说明、测试范围、精度影响，以及风险/兼容性说明后再提审。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/transformer/kimi_delta_attention.py (hunk 1 above)
Comment (Waynezee) [reply src/paddlefleet/transformer/kimi_delta_attention.py:94]: Done

**Review by From00 — APPROVED** [review]
LGTM

File: src/paddlefleet/transformer/kimi_delta_attention.py (hunk 1 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/kimi_delta_attention.py:94]: 复查当前 HEAD 后，这项只能判定为部分修复。新增方法已覆盖全部参数，但 `in_proj` 和 `conv1d` 的 TP 内容布局是按 q/k/v/beta/gate 各段分别切分后再拼接（新增多卡测试的 `_shard_param` 也按此处理）；当前却分别委托普通 `ColumnParallelLinear` 的连续 axis=1 规则和整块 axis=0 规则，因此各 rank 的 `global_offset` 不能表达这些非连续分段。现有测试仅核对 `global_shape`，TP=4 保存后按 TP=1 恢复仍会得到错误的 rank-major 排列。

请为这两个 fused 参数提供与分段布局一致的 checkpoint 表示，或拆成独立 q/k/v/... 参数，并补一个 TP=4 保存、TP=1 恢复后逐值对齐的测试；该线程暂不关闭。

File: src/paddlefleet/transformer/kimi_delta_attention.py [hunk 3]
```diff
@@ -0,0 +1,794 @@
...
+        query = query.reshape([batch, seq_len, -1, self.key_head_dim])
+        key = key.reshape([batch, seq_len, -1, self.key_head_dim])
+        value = value.reshape([batch, seq_len, -1, self.value_head_dim])
+
+        if self.use_qk_l2norm:
+            query = _l2norm(query.contiguous())
+            key = _l2norm(key.contiguous())
+
+        nvtx_range_push(suffix="kimi_delta_rule")
+        if (not HAVE_FLA) or self.config.deterministic_mode:
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/kimi_delta_attention.py:467]: P1 **优先级：P1**
**处理要求：请针对该评论修复并提交新的 commit。**

这里会让安装了 `fla` 的默认生产配置直接不可用：`fla.ops.kda` 导入成功时 `HAVE_FLA=True`，而 `deterministic_mode` 默认是 `False`，因此条件为假并在下面固定抛出 `NotImplementedError`。新增的单卡和多卡测试都显式设置了 `deterministic_mode=True`，没有覆盖这个默认分支。

在 FLA 调用真正接通前，请无条件使用 `paddle_chunk_kda`；或者在 `else` 中实际调用 `chunk_kda`。同时增加 `HAVE_FLA=True`、`deterministic_mode=False` 的回归测试，确保可选依赖的存在不会使 KDA 前向失败。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
复查当前 HEAD 后，checkpoint finding 仅部分修复：参数已经进入 sharded state dict，但 `in_proj` / `conv1d` 的分段 TP 布局仍未被 checkpoint offset 正确表达；此前的 mask finding 也仍然存在。另发现默认配置在 `fla` 可用时会直接抛出 `NotImplementedError`，详见新的 inline comment。

PR 正文现已补齐，最新的 `Check PR Description` 已通过；标题仍保留 `[WIP]`，正式提审前请改为规范 Tag。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>
